[pytorch_ci] Python target determinator

yns88 opened the pull request:
Summary:

This will make it so that if a pull request is just pure Python files, then we'll only run the Python tests that are connected to the dependency graph of the touched files.

Test Plan:
CircleCI results when touching just a few Python files:
- pytorch_macos_10_13_py3_test: 41 ->13 minutes https://circleci.com/gh/pytorch/pytorch/4550574?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_windows_vs2019_py36_cuda10.1_test1: 11 -> 2 minutes https://circleci.com/gh/pytorch/pytorch/4550846?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_windows_vs2019_py36_cuda10.1_test2: 51 -> 21 minutes https://circleci.com/gh/pytorch/pytorch/4550845?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_linux_xenial_py3_6_gcc5_4_test: 41 -> 14 minutes https://circleci.com/gh/pytorch/pytorch/4550543?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Differential Revision: D20009089

suo commented:
"test_X.py never imports from test_Y.py"

This is not always true, see test_jit.py

yns88 commented:
> "test_X.py never imports from test_Y.py"
> 
> This is not always true, see test_jit.py

Good catch. I'll need to deal with the FIXME on line 566, then -- the issue is that Modulefinder has difficulty traversing that graph unless the Python process is run from the "test/" path...

ezyang commented:
So... the big annoying question: how do we test that this is working correctly (that is to say, it is *not* improperly just throwing everything out from the to be tested set and thus we get all green?)

ezyang commented on test/run_test.py:
```diff
@@ -101,6 +103,32 @@
 ]
 
+# These tests are slow enough that it's worth calculating whether the patch
+# touched any related files first.
```
Is there a reason we don't just universally apply the determinator to everything?

ezyang replied:
Answer below appears to be, it is faster to run them than to determine them? (That sounds like an unusually fast test file...)

yns88 replied:
A lot of the test files run in less than 2 seconds!

ezyang replied:
If only we could collect stats on this.....

ezyang commented on test/run_test.py:
```diff
@@ -460,6 +599,14 @@ def main():
 
+    if options.determine_from is not None and os.path.exists(options.determine_from):
+        # HACK: Ensure the 'test' paths can be traversed by Modulefinder
+        sys.path.append('test')
```
I am... pretty sure this isn't going to work for nested tests, e.g., `test/distributed`

yns88 replied:
> Overall: This is neat, but it is a kind of chunky piece of code that, if done wrong, could lead to us missing CI signal on PRs. So it would be good to get a clearer characterization of what benefits we'll get when this gets merged.

I just ran an analysis over the most recent 100 master commits (I'm in the process of turning that into a unit test to assuage fears of under-testing). Out of the 21 "slow tests", this determination would have run a mean of 17 tests per commit. test_nn would run on ~80% of the commits. Determination will only run on PRs and not on master, but the makeup of master commits is probably similar to that of PR commits.

ezyang commented:
It would be helpful to have an English description of the algorithm somewhere

ezyang commented on test/run_test.py:
```diff
@@ -444,6 +475,114 @@ def get_selected_tests(options):
+    repo_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+    test_location = os.path.join(repo_root, 'test', test + '.py')
+    finder = modulefinder.ModuleFinder(
+        # Ideally exclude all third party modules, to speed up calculation.
```
Excludes seems like the wrong way to do this, on multiple accounts: (1) it appears that ModuleFinder is doing an exact match for entries in excludes, which means that if I directly import a second level module I'll zoom off into the netherworld of those modules as well, (2) a **whitelist** seems like a far more appropriate method of deciding whether or not to consider modules, since it's easy to say we only care about `torch` modules. But it doesn't see like ModuleFinder permits predicates, so you'll have to either monkeypatch the class or reimplement it yourself. I wonder if this is why ModuleFinder takes a long time; have you run it under verbose mode and looked at what it's traversing?

yns88 replied:
Another problem with a whitelist approach is that imports within the `test/` path do not share a common "test." namespace -- so you'd have to do some magic to even produce the whitelist.

ezyang commented:
In terms of impact, do you have any idea what percentage of submitted PRs would actually benefit from this determination? A simple test would be to collect the most recent N PRs and see what percentage of them only modify Python files.

ezyang commented:
Overall: This is neat, but it is a kind of chunky piece of code that, if done wrong, could lead to us missing CI signal on PRs. So it would be good to get a clearer characterization of what benefits we'll get when this gets merged.

yns88 commented:
> Overall: This is neat, but it is a kind of chunky piece of code that, if done wrong, could lead to us missing CI signal on PRs. So it would be good to get a clearer characterization of what benefits we'll get when this gets merged.

I just ran an analysis over the most recent 100 master commits (I'm in the process of turning that into a unit test to assuage fears of under-testing). Out of the 21 "slow tests", this determination would have run a mean of 17 tests per commit. test_nn would run on 81/100 commits. Determination will only run on PRs and not on master, but the makeup of master commits is probably similar to that of PR commits.

ezyang commented:
> I just ran an analysis over the most recent 100 master commits (I'm in the process of turning that into a unit test to assuage fears of under-testing). Out of the 21 "slow tests", this determination would have run a mean of 17 tests per commit. test_nn would run on 81/100 commits. Determination will only run on PRs and not on master, but the makeup of master commits is probably similar to that of PR commits.

Help me understand, are you saying that only 12 of these had only Python changes?

yns88 commented:
> Help me understand, are you saying that only 12 of these had only Python changes?

It turned out I had improperly installed the torch application on my machine when doing that. I just ran a more accurate calculation over the 500 most recent master commits:

- 110/500 commits touched only python files
- 119/500 commits touched only python files or non-code paths (docs/, README, etc)
- 136/500 commits touched only python files, non-code paths, or CI config paths
- 76/500 commits would skip test_nn (15.2%)
- The average number of "slow tests" determined on each commit is 22.8/27 (the median is 27 though, most commits would run either everything or nearly nothing)

However, if we decide that any change to `(.jenkins/|.circleci/|test/run_test.py)` should force all tests to run, the figures would bump up to 24.3/27 and 9.3% of commits skipping test_nn.

ezyang commented on .jenkins/pytorch/common.sh:
```diff
@@ -179,3 +179,11 @@ function get_exit_code() {
+  set +e
+  git fetch origin master --quiet
+  set -e
+  git diff --name-only "$(git merge-base origin master HEAD)" > "$1"
```
Is this always right? We have stack diffs and it seems like it might be inappropriate to compare against master, as you'll pick up all the old Python changes. Perhaps it's the case that we'll only over trigger in this case? But this might not be true if there are some reverts involved??

ezyang commented on test/run_test.py:
```diff
@@ -450,6 +489,135 @@ def get_selected_tests(options):
+        NONE - known to have no effect on test outcome
+        CI - CI configuration files
+    """
+    parts = filename.split(os.sep)
```
isn't `os.path.split(filename)` the more idiomatic way to do this?

yns88 replied:
Welp no, os.path.split is hardcoded to a maxsplit of 1 :/

ezyang commented on test/run_test.py:
```diff
@@ -349,6 +385,9 @@ def parse_args():
         help='always run blacklisted windows tests')
+    parser.add_argument(
+        '--determine-from',
+        help='File of affected source filenames to determine which tests to run.')
```
Should determination be exposed as a developer mode they can toggle directly? The workflow would be something like, "I change  a Python file, now I want to run every test that might have been affected by that Python file."

ezyang commented on test/run_test.py:
```diff
@@ -450,6 +489,135 @@ def get_selected_tests(options):
+        finder.run_script(test_location)
+    else:
+        with open(test_location, 'r', encoding='utf-8') as fp:
+            stuff = ('', 'r', 1)
```
stuff? :o)

yns88 replied:
Indeed, the canonical arg name 😄https://github.com/python/cpython/blob/3.8/Lib/modulefinder.py#L164

ezyang commented on test/run_test.py:
```diff
@@ -450,6 +489,135 @@ def get_selected_tests(options):
+            continue
+        elif file_type == 'CI':
+            # Run the determination tests on a CI change, but not other tests.
+            # FIXME: In the future, run all tests on a CI change.
```
Shouldn't you fix this FIXME before putting this in? In general if I change the CI scripts I can modify environment variables or arbitrary other stuff, and I'd better run everything in that case.

ezyang commented:
The determinator tests are a good start. There are a few more cases which are likely to perturb the determinator, and it would be good to make sure they get handled well:

1. If someone adds a new Python module folder, or moves Python modules around, what happens? (I think this is OK)
2. If someone adds a new test script, what happens? (I think this OK; but I'm still not sure that tests in nested directories are handled correctly)
3. If a previously fast running test becomes slow (because people have added a lot of tests to it), what happens? (I think we don't notice. Hard to notice without stats.)
4. If someone adds a new top-level folder to the PyTorch repository (e.g., think of mobile), what happens? (I think this is OK, it gets classified as UNKNOWN and we always run)
5. If someone adds a new third-party dependency, what happens? (Determinator runs more slowly, because that dependency is not blacklisted. Also hard to notice without stats.)

Also, it would be really bad if we land this and due to a bug in the PR versus master logic we keep running the determinator. We should check for this when it lands.

ezyang reviewed (approved):
Willing to give this a shot, but I'll feel more comfortable if we have some way of evaluating how well things go. For example, if we print out determinator results to CI log and then parse them out in Dr. CI we would be able to monitor what's going on with the determinator.

yns88 committed:
[pytorch_ci] Python target determinator (#33577)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/33577

Pull Request resolved: https://github.com/pytorch/pytorch/pull/33221

This will make it so that if a pull request is just pure Python files, then we'll only run the Python tests that are connected to the dependency graph of the touched files.

Assumptions made:
- the Python code does not do dynamic imports
- test_X.py never imports from test_Y.py

Right now this is only done for test_nn (presumably the largest test entrypoint), but it's not much more work to do it for all the other test entrypoints too.

Test Plan:
CircleCI results when touching just a few Python files:
- pytorch_macos_10_13_py3_test: 41 ->13 minutes https://circleci.com/gh/pytorch/pytorch/4550574?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_windows_vs2019_py36_cuda10.1_test1: 11 -> 2 minutes https://circleci.com/gh/pytorch/pytorch/4550846?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_windows_vs2019_py36_cuda10.1_test2: 51 -> 21 minutes https://circleci.com/gh/pytorch/pytorch/4550845?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
- pytorch_linux_xenial_py3_6_gcc5_4_test: 41 -> 14 minutes https://circleci.com/gh/pytorch/pytorch/4550543?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Differential Revision: D20009089

fbshipit-source-id: 71dbaabda5c6b7325ac418b0aef2f780221a3a2c

linbinyu commented:
seems break multiple CI tests like https://circleci.com/gh/pytorch/pytorch/4686169

yns88 commented:
If you're referring to the 403 error in download_mnist.py, that appears to have started a few hours before this merged.

mrshenli commented:
This test error on master seems to start from this PR, but looks like it passed in the PR test. 

```
Mar 04 08:46:14 ======================================================================
Mar 04 08:46:14 ERROR: test_tanh (__main__.TestTorchFunctionOverride)
Mar 04 08:46:14 ----------------------------------------------------------------------
Mar 04 08:46:14 Traceback (most recent call last):
Mar 04 08:46:14   File "test_overrides.py", line 478, in test
Mar 04 08:46:14     self.assertEqual(func(*func_args), -1)
Mar 04 08:46:14   File "/opt/python/3.5/lib/python3.5/site-packages/torch/nn/functional.py", line 1558, in tanh
Mar 04 08:46:14     return input.tanh()
Mar 04 08:46:14 AttributeError: 'TensorLike' object has no attribute 'tanh'
```

mrshenli commented:
Oh, those are actually flaky

https://app.circleci.com/jobs/github/pytorch/pytorch/4685328

mrshenli commented:
The new tests break master:

https://app.circleci.com/jobs/github/pytorch/pytorch/4692476
https://app.circleci.com/jobs/github/pytorch/pytorch/4692633

I am reverting it for now to make master green. Please resubmit when the error is fixed. Thanks!

```
Mar 04 20:31:12 ======================================================================
Mar 04 20:31:12 ERROR [5.426s]: test_caffe2_file (__main__.DeterminationTest)
Mar 04 20:31:12 Caffe2 files trigger dependent tests
Mar 04 20:31:12 ----------------------------------------------------------------------
Mar 04 20:31:12 Traceback (most recent call last):
Mar 04 20:31:12   File "test_determination.py", line 103, in test_caffe2_file
Mar 04 20:31:12     self.assertEqual(self.determined_tests(["caffe2/python/brew_test.py"]), [])
Mar 04 20:31:12   File "test_determination.py", line 32, in determined_tests
Mar 04 20:31:12     for test in cls.TESTS
Mar 04 20:31:12   File "test_determination.py", line 33, in <listcomp>
Mar 04 20:31:12     if run_test.determine_target(test, changed_files, DummyOptions())
Mar 04 20:31:12   File "/var/lib/jenkins/workspace/test/run_test.py", line 584, in determine_target
Mar 04 20:31:12     dep_modules = get_dep_modules(test)
Mar 04 20:31:12   File "/var/lib/jenkins/workspace/test/run_test.py", line 564, in get_dep_modules
Mar 04 20:31:12     finder.load_module('__main__', fp, test_location, ('', 'r', 1))
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 304, in load_module
Mar 04 20:31:12     self.scan_code(co, m)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 375, in scan_code
Mar 04 20:31:12     self._safe_import_hook(name, m, fromlist, level=0)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 322, in _safe_import_hook
Mar 04 20:31:12     self.import_hook(name, caller, level=level)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 122, in import_hook
Mar 04 20:31:12     q, tail = self.find_head_package(parent, name)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 178, in find_head_package
Mar 04 20:31:12     q = self.import_module(head, qname, parent)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 271, in import_module
Mar 04 20:31:12     m = self.load_module(fqname, fp, pathname, stuff)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 304, in load_module
Mar 04 20:31:12     self.scan_code(co, m)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 375, in scan_code
Mar 04 20:31:12     self._safe_import_hook(name, m, fromlist, level=0)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 322, in _safe_import_hook
Mar 04 20:31:12     self.import_hook(name, caller, level=level)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 122, in import_hook
Mar 04 20:31:12     q, tail = self.find_head_package(parent, name)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 178, in find_head_package
Mar 04 20:31:12     q = self.import_module(head, qname, parent)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 271, in import_module
Mar 04 20:31:12     m = self.load_module(fqname, fp, pathname, stuff)
Mar 04 20:31:12   File "/opt/python/3.5/lib/python3.5/modulefinder.py", line 288, in load_module
Mar 04 20:31:12     co = compile(fp.read()+'\n', pathname, 'exec')
Mar 04 20:31:12   File "/var/lib/jenkins/workspace/test/test_jit_py3.py", line 17
Mar 04 20:31:12     print(f"{hello + ' ' + test}, I'm a {test}") # noqa E999
Mar 04 20:31:12                                               ^
Mar 04 20:31:12 SyntaxError: invalid syntax
```
